utilities/cli/util.py

```diff
@@ -319,6 +391,61 @@ def __init__(self, package_name: str, version_pattern: str, message: str = None)
         )
 
 
+def is_package_installed(package_name: str) -> bool:
+    """Check if a package is already installed"""
+    try:
+        result = subprocess.run(
+            ["dpkg", "-l", package_name], capture_output=True, text=True, check=False
+        )
+        return result.returncode == 0 and "ii" in result.stdout
+    except (subprocess.CalledProcessError, FileNotFoundError):
+        return False
+
+
+def ensure_apt_updated(dry_run: bool = False) -> None:
+    """Ensure apt package list is updated, but only once per session"""
+    global _apt_updated
+    if not _apt_updated:
+        run_command(["apt-get", "update"], dry_run=dry_run)
+        _apt_updated = True
+
+
+def install_packages_if_missing(
+    packages: List[str], dry_run: bool = False, apt_options: List[str] = None
+) -> List[str]:
+    """Install packages only if they're not already installed
+
+    Args:
+        packages: List of package names to install (can include version specs like "pkg=1.0*")
+            Note: If package has a version spec, it will run [sudo] apt install to ensure version.
+        dry_run: Whether to perform a dry run
+        apt_options: Additional options for apt install (e.g., ["--no-install-recommends", "-y"])
+
+    Returns:
+        List of packages that were actually installed (or would be installed in dry run)
+    """
+    if apt_options is None:
+        apt_options = ["--no-install-recommends", "-y"]
+
+    missing_packages = []
+    for package in packages:
+        package_name = package.split("=")[0]
+        if "=" in package:
+            missing_packages.append(package)
+            info(f"Package {package_name} will be checked/installed with version specification")
+        elif not is_package_installed(package_name):
+            missing_packages.append(package)
+        else:
+            info(f"Package {package_name} is already installed")
+
+    if missing_packages:
+        ensure_apt_updated(dry_run=dry_run)
+        install_cmd = ["apt", "install"] + apt_options + missing_packages
+        run_command(install_cmd, dry_run=dry_run)
```

Comment:
What is the behavior if any of the packages are not found in the apt cache?

Reply:
in that case the entire process will exit after the `apt install`. I've also updated the `handle_setup` to make it clear that there are these stages:
```
        holohub_cli_util.install_packages_if_missing(
            ["wget", "xvfb", "git", "unzip", "ffmpeg", "ninja-build", "libv4l-dev"],
            dry_run=args.dryrun,
        )
        holohub_cli_util.setup_cmake(dry_run=args.dryrun)
        holohub_cli_util.setup_python_dev(dry_run=args.dryrun)
        holohub_cli_util.setup_ngc_cli(dry_run=args.dryrun)
        holohub_cli_util.setup_cuda_dependencies(dry_run=args.dryrun)
```
any error within the stage will exit the process.

didn't add any retry logic here as it'll be more intuitive to manually run `./holohub setup` again if needed